src/mon/OSDMonitor.cc

```diff
@@ -6993,32 +6993,145 @@ int OSDMonitor::prepare_new_pool(string& name,
+{
+  dout(10) << __func__ << " flag " << name << " for "
+	   << (is_set ? "set" : "unset") << " sure? " << is_sure << dendl;
+
+  if (flag == CEPH_OSDMAP_PGLOG_HARDLIMIT) {
+    if (!is_set) {
+      ss << "flag '" << name << "' cannot be unset";
+      return -EINVAL;
+    }
+    bool needs_is_sure = g_conf().get_val<bool>(
+	"mon_debug_osd_force_is_sure");
+    if ((needs_is_sure || !osdmap.get_num_up_osds()) && !is_sure) {
+      ss << "Not advisable to continue since no OSDs are up. Pass "
+	 << "--yes-i-really-mean-it if you really wish to continue.";
+      return -EPERM;
+    }
+    if (osdmap.require_osd_release < CEPH_RELEASE_LUMINOUS ||
+	(!HAVE_FEATURE(osdmap.get_up_osd_features(), OSD_PGLOG_HARDLIMIT)
+	 && !is_sure)) {
+      ss << "not all up OSDs have OSD_PGLOG_HARDLIMIT feature";
+      return -EPERM;
+    }
+  }
+  return 0;
+}
+
+bool OSDMonitor::_do_set_unset_flags(
+    MonOpRequestRef op,
+    uint64_t flags,
+    bool is_set)
 {
-  op->mark_osdmon_event(__func__);
-  ostringstream ss;
-  if (pending_inc.new_flags < 0)
+  string flag_str = osdmap.get_flag_string(flags);
+  if (flag_str.empty()) {
+    return false;
+  }
+
+  if (pending_inc.new_flags < 0) {
     pending_inc.new_flags = osdmap.get_flags();
-  pending_inc.new_flags |= flag;
-  ss << OSDMap::get_flag_string(flag) << " is set";
-  wait_for_finished_proposal(op, new Monitor::C_Command(mon, op, 0, ss.str(),
-						    get_last_committed() + 1));
+  }
+
+  string ret_str;
+  if (is_set) {
+    pending_inc.new_flags |= flags;
+    ret_str = flag_str + " is set";
+  } else {
+    pending_inc.new_flags &= ~flags;
+    ret_str = flag_str + " is unset";
+  }
+
+  dout(10) << __func__ << " queue pending: " << ret_str << dendl;
+
+  wait_for_finished_proposal(
+      op,
+      new Monitor::C_Command(mon, op, 0, ret_str, get_last_committed() + 1));
   return true;
 }
 
-bool OSDMonitor::prepare_unset_flag(MonOpRequestRef op, int flag)
+bool OSDMonitor::prepare_set_flag(
+    MonOpRequestRef op,
+    uint64_t flags)
 {
-  op->mark_osdmon_event(__func__);
-  ostringstream ss;
-  if (pending_inc.new_flags < 0)
-    pending_inc.new_flags = osdmap.get_flags();
-  pending_inc.new_flags &= ~flag;
-  ss << OSDMap::get_flag_string(flag) << " is unset";
-  wait_for_finished_proposal(op, new Monitor::C_Command(mon, op, 0, ss.str(),
-						    get_last_committed() + 1));
+  _do_set_unset_flags(op, flags, true);
   return true;
 }
 
+bool OSDMonitor::prepare_unset_flag(
+    MonOpRequestRef op,
+    uint64_t flags)
+{
+  stringstream ss;
+  _do_set_unset_flags(op, flags, false);
+  return true;
+}
+
+int OSDMonitor::prepare_set_unset_flags(
+    MonOpRequestRef op,
+    vector<string> flags,
+    bool is_set,
+    bool sure,
+    stringstream &ss)
+{
+  dout(10) << __func__ << " " << (is_set ? "set" : "unset" )
+	   << " " << flags << " sure? " << sure << dendl;
+
+  // essentially, key name -> (flag, requires special handling)
+  map<string, tuple<uint64_t, bool> > flag_map = {
```

Comment:
Sorry for not catching this earlier, but since this is at function scope you may want to mark it "static" so that flag_map is not initialized on each call. (I don't expect these are particularly performance-sensitive, but it's easy enough to do.)

Reply:
Yeah, if I had a time machine I might not use a boolean here, and use an enum such as flag_special_handler. I like tuples, but they can be a slippery slope. I can go either way on that, but I tend to prefer more explicitness in these cases. Does the comment about "static" make sense?